### PR TITLE
update 1.6

### PR DESCRIPTION
**Pull request by kndnsow**
rebuild ajax with external service function
removed Facebook share
clean up with lib functions

### CODE REVIEW

File: amd/src/analytic.js [hunk 1]
```diff
@@ -0,0 +1,480 @@
+define(['jquery', 'core/ajax'], function ($, ajax) {
+    return {
+        analytic: function (user_id, course_id) {
+
+
+            $(document).ready(function () {
```
Comment (purnendudash) [inline amd/src/analytic.js]: Should this be here?

File: amd/src/analytic.js [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+                var lastattemptsummary, loggedinuser, mixchart, allusers, questionpercat, timechart, gradeanalysis, quesanalysis, hardestques, allquestions, quizid, rooturl, userid, lastuserquizattemptid;
+                var attemptssnapshot_arr = [];
+                Chart.plugins.register({
+                    beforeDraw: function (chartInstance) {
+                        var ctx = chartInstance.chart.ctx;
+                        ctx.fillStyle = "white";
+                        ctx.fillRect(0, 0, chartInstance.chart.width, chartInstance.chart.height);
+                    }
+                });
+
```
Comment (purnendudash) [inline amd/src/analytic.js]: Why fetchdata?

File: amd/src/analytic.js [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+                                },
+                            };
+                            var questionpercatopt = $.extend(totaldata.questionpercat.opt, questionpercatopt2);
+
+                            questionpercat = new Chart(ctx, {
+                                type: 'pie',
+                                data: totaldata.questionpercat.data,
+                                options: questionpercatopt,
+                            });
+                            document.getElementById('js-legendqpc').innerHTML = questionpercat.generateLegend();
```
Comment (purnendudash) [inline amd/src/analytic.js]: Why normal js?

File: amd/src/analytic.js [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+                                type: 'bar',
+                                data: totaldata.hardestques.data,
+                                options: hardestquesopt
+                            });
+                        }
+
+                    }).fail(function (ex) {
+
+                    });
+                    var canvasquesanalysis = document.getElementById("quesanalysis");
```
Comment (purnendudash) [inline amd/src/analytic.js]: replace all such getElementById by some query

File: externallib.php [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+    }
+
+
+    public static function quizanalytics_fetchdata($quizid)
+    {
+
+        global $DB, $USER, $PAGE, $CFG;
+
+
+        // $params = self::validate_parameters(
```
Comment (purnendudash) [inline externallib.php]: Please clean all such comments thing

File: index.php [hunk 6]
```diff
@@ -127,187 +102,489 @@
...
-        </div>';
+                </div>';
 
 echo $html;
 
-echo $OUTPUT->footer();
\ No newline at end of file
+echo $OUTPUT->footer();
+?>
+<style>
```
Comment (purnendudash) [inline index.php:260]: Why so many inline css

File: thirdpartylibs.xml [hunk 7]
```diff
@@ -1,51 +0,0 @@
-<?xml version="1.0"?>
```
Comment (purnendudash) [inline thirdpartylibs.xml:1]: I guess this file should be there.

File: amd/src/analytic.js [hunk 8]
```diff
@@ -0,0 +1,477 @@
+define(['jquery', 'core/ajax'], function ($, ajax) {
+    return {
+        analytic: function (user_id, course_id) {
+
+            var lastattemptsummary, loggedinuser, mixchart, allusers, questionpercat, timechart, gradeanalysis, quesanalysis, hardestques, allquestions, quizid, rooturl, userid, lastuserquizattemptid;
```
Comment (cool-raju) [inline amd/src/analytic.js]: please check the Naming Conventions all over the code

File: amd/src/analytic.js [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+                beforeDraw: function (chartInstance) {
+                    var ctx = chartInstance.chart.ctx;
+                    ctx.fillStyle = "white";
+                    ctx.fillRect(0, 0, chartInstance.chart.width, chartInstance.chart.height);
+                }
+            });
+
+            $(".viewanalytic").click(function (e) {
+                e.preventDefault();
+
```
Comment (cool-raju) [inline amd/src/analytic.js]: please remove all the extra spaces.

File: amd/src/analytic.js [hunk 10]
```diff
@@ -0,0 +1,477 @@
...
+                var quizid = $(this).data('quize_id');
+
+                var promises = ajax.call([
+                    {
+                        methodname: 'moodle_quizanalytics_analytic',
+                        args: { quizid: quizid },
+                    }
+                ]);
+                promises[0].done(function (data) {
```
Comment (cool-raju) [inline amd/src/analytic.js:21]: add checking before directly using an array data

File: amd/src/analytic.js [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+                    if (data) {
+                        var totaldata = jQuery.parseJSON(data);
+
+                        allquestions = totaldata.allquestion;
```
Comment (cool-raju) [inline amd/src/analytic.js]: what happen if totaldata give nulled value???check all the related

File: amd/src/analytic.js [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+                            hardestques.destroy();
+                        }
+                        hardestques = new Chart(ctx, {
+                            type: 'bar',
+                            data: totaldata.hardestques.data,
+                            options: hardestquesopt
+                        });
+                    }
+
+                }).fail(function (ex) {
```
Comment (cool-raju) [inline amd/src/analytic.js]: why empty call?? add a log or remove it

File: amd/src/analytic.js [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+
+                });
+                var canvasquesanalysis = document.getElementById("quesanalysis");
+                var canvashardestques = document.getElementById("hardestques");
+
+                canvasquesanalysis.onclick = function (qevt) {
+                    var activePoints = quesanalysis.getElementsAtEvent(qevt);
```
Comment (cool-raju) [inline amd/src/analytic.js]: check every variables before adding amy method on it

File: amd/src/analytic.js [hunk 14]
```diff
@@ -1,7 +1,7 @@
 define(['jquery', 'core/ajax'], function ($, ajax) {
     return {
-        analytic: function (user_id, course_id) {
-            var last_attempt_summary, loggedin_user, mixchart, all_users, questionpercat, timechart, gradeanalysis, quesanalysis, hardestques, all_questions, quiz_id, rooturl, user_id, lastuserquizattemptid;
+        analytic: function () {
+            var lastAttemptSummary, loggedInUser, mixChart, allUsers, questionPerCat, timeChart, gradeAnalysis, quesAnalysis, hardestques, allQuestions, quizid, rooturl, userid, lastUserQuizAttemptID;
             var attemptssnapshot_arr = [];
```
Comment (cool-raju) [inline amd/src/analytic.js]: give proper logical name

File: externallib.php [hunk 15]
```diff
@@ -0,0 +1,688 @@
...
+    {
+        return new external_function_parameters(
+            array(
+                'quizid' => new external_value(PARAM_INT, 'quiz id')
+            )
+        );
+    }
+    public static function quizanalytics_analytic($quizid)
+    {
+        global $DB, $USER, $PAGE, $CFG;
```
Comment (cool-raju) [inline externallib.php]: check whether all global define variables has used or not

File: externallib.php [hunk 16]
```diff
@@ -0,0 +1,688 @@
...
+        $quiz = $DB->get_record('quiz', array('id' => $quizid));
+
+        $attemptssql = "SELECT * FROM {quiz_attempts}
+          WHERE state = 'finished' AND sumgrades IS NOT NULL AND quiz = ?";
+
+        $totalquizattempted = $DB->get_records_sql($attemptssql, array($quizid));
```
Comment (cool-raju) [inline externallib.php]: is this a proper name??.. naming convention ?? check all d related

File: externallib.php [hunk 17]
```diff
@@ -0,0 +1,688 @@
...
+                qs.quizid = ? AND q.qtype != ? GROUP BY qc.id", array($quizid, 'description'));
+            $catname = array();
+            $catdata = array();
+            $randomcatcolor = array();
+            $overallhardness = array();
+            $loggdinuserhardness = array();
+            $totaluserswrongattemts = array();
+            $totalwrongattemts = array();
+            foreach ($catdetails as $catdetail) {
+                $catname[] = $catdetail->name;
```
Comment (cool-raju) [inline externallib.php]: before fetching any object or array value checking is compulsory..check for all

File: externallib.php [hunk 18]
```diff
@@ -0,0 +1,688 @@
...
+             */
+            function random_color()
+            {
+                return random_color_part() . random_color_part() . random_color_part();
+            }
+            $catdetails = $DB->get_records_sql("SELECT qc.id, COUNT(qs.id) as qnum,
+                qc.name FROM {quiz_slots} qs, {question} q, {question_categories} qc, {question_bank_entries} qbe
+                WHERE q.id = qs.id AND qbe.questioncategoryid = qc.id AND
+                qs.quizid = ? AND q.qtype != ? GROUP BY qc.id", array($quizid, 'description'));
+            $catname = array();
```
Comment (cool-raju) [inline externallib.php]: 47 to 53..why all these arrays.. define multiple value blank arrays on a single line

File: externallib.php [hunk 19]
```diff
@@ -0,0 +1,688 @@
...
+            $questionpercatopt = array('legend' => array(
+                'display' => false,
+                'position' => 'bottom', 'labels' => array('boxWidth' => 13)
+            ), 'title' => array(
+                'display' => true,
+                'position' => 'bottom', 'text' => get_string('questionspercategory', 'gradereport_quizanalytics')
+            ));
+            /* allusers */
+            arsort($overallhardness);
+            $maxhardnesskeys = array_keys($overallhardness, max($overallhardness));
```
Comment (cool-raju) [inline externallib.php]: use the ternary operator to checking values before inserting any funtion

File: externallib.php [hunk 20]
```diff
@@ -0,0 +1,688 @@
...
+                'position' => 'bottom'
+            ), 'title' => array(
+                'display' => false,
+                'position' => 'bottom', 'text' => get_string('hardcatalluser', 'gradereport_quizanalytics')
+            ));
+            /* loggedinuser */
+            arsort($loggdinuserhardness);
+            $maxloggdinuserhardnesskeys = array_keys($loggdinuserhardness, max($loggdinuserhardness));
+
+            foreach ($maxloggdinuserhardnesskeys as $maxloggdinuserhardnesskey) {
```
Comment (cool-raju) [inline externallib.php]: give a specific/understandable name... already explain multiple times.. this is d last time mentioned here

File: amd/src/analytic.js [hunk 21]
```diff
@@ -28,7 +28,7 @@ define(['jquery', 'core/ajax'], function ($, ajax) {
                         $(".showanalytics").find(".parentTabs").find("span.lastattemptsummary").hide();
                         $(".showanalytics").find("#tabs-1").find("p.lastattemptsummarydes").hide();
                         $(".showanalytics").find("#tabs-1").find("p.attemptsummarydes").show();
-                        if (totalData.userattempts > 1) {
```
Comment (cool-raju) [inline amd/src/analytic.js]: check every value for here "totalData" before use it further

**Pull request merged**